src/array/cuda/gather_mm.cu

```diff
@@ -0,0 +1,334 @@
+
+            DType* A_trans_data = nullptr;
+            DType* B_trans_data = nullptr;
+            if (a_trans) {
+                A_trans_data = static_cast<DType*>(device->AllocWorkspace \
+                    (A->ctx, m * k * sizeof(DType)));
+                _Transpose(thr_entry->cublas_handle, A_data + A_offset, A_trans_data, m, k);
+            }
+            if (b_trans) {
+                B_trans_data = static_cast<DType*>(device->AllocWorkspace \
+                    (B->ctx, n * k * sizeof(DType)));
+                _Transpose(thr_entry->cublas_handle, B_data + B_offset, B_trans_data, k, n);
+            }
+            if (a_trans || b_trans) {
+                int64_t tmp = k;
+                if (a_trans)
+                    std::swap(m, k);
+                if (b_trans)  {
+                    k = tmp;
+                    std::swap(n, k);
+                }
+            }
+            int ldb = n, lda = k, ldc = n;
+            CUBLAS_CALL(cublasGemm<DType>(
+                thr_entry->cublas_handle,
+                CUBLAS_OP_N,
+                CUBLAS_OP_N,
+                n, m, k,
+                &alpha,
+                (b_trans) ? B_trans_data : B_data + B_offset, n,
+                (a_trans) ? A_trans_data : A_data + A_offset, k,
+                &beta,
+                C_data + C_offset, n));
+            if (a_trans)
+                device->FreeWorkspace(A->ctx, A_trans_data);
+            if (b_trans)
+                device->FreeWorkspace(B->ctx, B_trans_data);
+            A_offset += m * k;
+            B_offset += k * n;
+            C_offset += m * n;
+        }
+    });
+}
+
+/*!
+ * \brief Implementation of Gather_mm operator. The input matrix A is
+ *        expected to be sorted according to relation type.
+ * \param A The input dense matrix of dimension m x k
+ * \param B The input dense matrix of dimension k x n
+ * \param C The output dense matrix od dimension m x n
+ * \param A_dim1_per_rel The number of rows in each relation in A
+ * \param B_dim1_per_rel The number of rows in each relation in B
+ * \param etype relation types of each edge. Required by *UnsortedEtype kernel
+ * \param sortedA Matrix A is sorted according to relation type or not
+ * \param a_trans Matrix A to be transposed
+ * \param b_trans Matrix B to be transposed
+ */
+template <int XPU, typename IdType, int bits>
+void gatherMM(const NDArray A,
+          const NDArray B,
+          NDArray C,
+          const NDArray A_dim1_per_rel,
+          const NDArray B_dim1_per_rel,
+          const NDArray etype,
+          bool sortedA, bool a_trans, bool b_trans) {
+    if (sortedA) {  // similar to low-mem matmul
+        gatherMM_SortedEtype<XPU, IdType, bits>(A, B, C, A_dim1_per_rel,
+            B_dim1_per_rel, a_trans, b_trans);
+    } else {  // similar to bmm (high-mem) without copying weights to edges
+        // TODO(Israt): Add support for A and B to have different dimension in relation types
```

Comment:
Does this need an assert/check?

Reply:
Yes. Added a check inside the called function.